### PR TITLE
[roslaunch] Add yaml type for <param> tag

### PR DESCRIPTION
**Pull request by wkentaro**
## Why?

To set dict, list string as a parameter.
This is ok if we use `<rosparam>` tag, but it does not have the feature of running arbitrary commands which `<param>` tag has.

## Example

My motivation is ways to use like below:

```bash
% cat apc2015_00.json
{
  "bin_contents":
  {
    "bin_I":
    [
      "sharpie_accent_tank_style_highlighters",
      "dr_browns_bottle_brush"
    ]
  },

  "work_order":
  [
    {
      "bin": "bin_I",
      "item": "sharpie_accent_tank_style_highlighters"
    }
  ]
}

% cat spam.launch
<launch>

  <param name="/test" type="yaml"
         command="/usr/local/bin/json2yaml $(find roslaunch)/apc2015_00.json" />

</launch>

% roslaunch ./spam.launch

% rosparam get /test
bin_contents:
  bin_I: [sharpie_accent_tank_style_highlighters, dr_browns_bottle_brush]
work_order:
- {bin: bin_I, item: sharpie_accent_tank_style_highlighters}
```

### CODE REVIEW

**Comment by mikepurvis** [comment]
I dig it. I've wanted something like this for ages.

File: tools/roslaunch/package.xml [hunk 1]
```diff
@@ -32,6 +32,7 @@
   <run_depend version_gte="1.13.3">rosunit</run_depend>
 
   <test_depend>rosbuild</test_depend>
+  <test_depend>rospy</test_depend>
```
Comment (dirk-thomas) [inline tools/roslaunch/package.xml]: Why is this necessary? I don't see any part in the patch which requires it.

File: tools/roslaunch/src/roslaunch/loader.py [hunk 2]
```diff
@@ -70,30 +70,36 @@ def convert_value(value, type_):
         #attempt numeric conversion
         try:
             if '.' in value:
-                return float(value)
+                return float(value.strip())
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/loader.py]: This is not necessary. `float()` already works even if the passed argument has leading / trailing spaces.

Same below.

File: tools/roslaunch/src/roslaunch/loader.py [hunk 3]
```diff
@@ -70,30 +70,36 @@ def convert_value(value, type_):
...
             else:
-                return int(value)
+                return int(value.strip())
         except ValueError as e:
             pass
         #bool
-        lval = value.lower()
+        lval = value.strip().lower()
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/loader.py]: In which case do you expect the strings `true` and `false` to have leading / trailing spaces?

File: tools/roslaunch/src/roslaunch/loader.py [hunk 4]
```diff
@@ -70,30 +70,36 @@ def convert_value(value, type_):
...
         if value == 'true' or value == '1':
             return True
         elif value == 'false' or value == '0':
             return False
         raise ValueError("%s is not a '%s' type"%(value, type_))
+    elif type_ == 'yaml':
+        # - lazy import
+        global yaml
+        if yaml is None:
+            import yaml
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/loader.py]: Please move the import to the top of the file and avoid the conditional block.

File: tools/roslaunch/src/roslaunch/loader.py [hunk 5]
```diff
@@ -70,30 +70,36 @@ def convert_value(value, type_):
...
+        return yaml.load(value)
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/loader.py]: How is that represented on the parameter server? Does this store a complex type (e.g. a nested dictionary) under a single `name`?

Comment (wkentaro) [reply tools/roslaunch/src/roslaunch/loader.py]: It is set like this:

```
% rosparam get /test1
bin_contents:
  bin_I: [sharpie_accent_tank_style_highlighters, dr_browns_bottle_brush]
work_order:
- {bin: bin_I, item: sharpie_accent_tank_style_highlighters}

% rosparam get /test1/bin_contents/bin_I
[sharpie_accent_tank_style_highlighters, dr_browns_bottle_brush]


% rosparam get /test1/bin_contents
bin_I: [sharpie_accent_tank_style_highlighters, dr_browns_bottle_brush]

% rosparam get /test1/bin_contents/bin_I
[sharpie_accent_tank_style_highlighters, dr_browns_bottle_brush]
```

File: tools/roslaunch/src/roslaunch/loader.py [hunk 6]
```diff
@@ -94,6 +95,8 @@ def convert_value(value, type_):
         elif value == 'false' or value == '0':
             return False
         raise ValueError("%s is not a '%s' type"%(value, type_))
+    elif type_ == 'yaml':
+        return yaml.load(value)
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/loader.py]: Should any yaml specific exceptions be catched here and a `ValueError` be raised instead?

**Comment by wkentaro** [comment]
Had error below:
```
% nosetests test_xmlloader.py
.............................E............
======================================================================
ERROR: test_params (unit.test_xmlloader.TestXmlLoader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/test/unit/test_xmlloader.py", line 192, in test_params
    mock = self._load(os.path.join(self.xml_dir, 'test-params-valid.xml'))
  File "/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/test/unit/test_xmlloader.py", line 100, in _load
    loader.load(test_file, mock)
  File "/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/src/roslaunch/xmlloader.py", line 746, in load
    self._load_launch(launch, ros_config, is_core=core, filename=filename, argv=argv, verbose=verbose)
  File "/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/src/roslaunch/xmlloader.py", line 718, in _load_launch
    self._recurse_load(ros_config, launch.childNodes, self.root_context, None, is_core, verbose)
  File "/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/src/roslaunch/xmlloader.py", line 662, in _recurse_load
    self._param_tag(tag, context, ros_config, verbose=verbose)
  File "/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/src/roslaunch/xmlloader.py", line 95, in call
    return f(*args, **kwds)
  File "/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/src/roslaunch/xmlloader.py", line 279, in _param_tag
    "Invalid <param> tag: %s. \n\nParam xml is %s"%(e, tag.toxml()))
XmlParseException: Invalid <param> tag: true
 is not a 'bool' type.

Param xml is <param command="echo true" name="commandoutputbool" type="bool"/>
-------------------- >> begin captured stdout << ---------------------
... loading XML file [/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/test/xml/test-params-valid.xml]
... executing command param [cat "/home/wkentaro/Projects/ros_comm_ws/src/ros/ros_comm/tools/roslaunch/resources/example.launch"]
... executing command param [echo 0]
... executing command param [echo 10]
... executing command param [echo true]

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 42 tests in 0.644s

FAILED (errors=1)
```

**Comment by dirk-thomas** [comment]
I am sorry for the late response. This looks good now. Merging...

Can you please add appropriate documentation for this new feature to the wiki and link to the change here. Thanks.

**Comment by wkentaro** [comment]
http://wiki.ros.org/action/diff/roslaunch/XML/param?action=diff

**Comment by dirk-thomas** [comment]
Thank you. I reformated the paragraph a little bit and changed the note that it is only available as of Lunar atm: http://wiki.ros.org/action/info/roslaunch/XML/param?action=diff&rev2=12&rev1=11
